### PR TITLE
Update Lua to 5.4.2

### PR DESCRIPTION
**Pull request by fresholia**
The Lua version selector, which is highly demanded by the developers, will add a great impact to the performance. I would appreciate if you comment my mistakes or thoughts.

### **ISSUES**
- Not being able to run two versions at the same time in lua parser files

_Sources_
Lua 5.4 source: https://www.lua.org/ftp/lua-5.4.2.tar.gz
LuaJIT source: https://luajit.org/download/LuaJIT-2.1.0-beta3.zip

### CODE REVIEW

**Comment by cleopatradev** [comment]
Edit the following for build on the local:

Lua_Client.vcxproj:
```xml
  <ItemGroup>
    <ClInclude Include="..\vendor\lua\src\lapi.h" />
    <ClInclude Include="..\vendor\lua\src\lauxlib.h" />
    <ClInclude Include="..\vendor\lua\src\lcode.h" />
    <ClInclude Include="..\vendor\lua\src\lctype.h" />
    <ClInclude Include="..\vendor\lua\src\ldebug.h" />
    <ClInclude Include="..\vendor\lua\src\ldo.h" />
    <ClInclude Include="..\vendor\lua\src\lfunc.h" />
    <ClInclude Include="..\vendor\lua\src\lgc.h" />
    <ClInclude Include="..\vendor\lua\src\ljumptab.h" />
    <ClInclude Include="..\vendor\lua\src\llex.h" />
    <ClInclude Include="..\vendor\lua\src\llimits.h" />
    <ClInclude Include="..\vendor\lua\src\lmem.h" />
    <ClInclude Include="..\vendor\lua\src\lobject.h" />
    <ClInclude Include="..\vendor\lua\src\lopcodes.h" />
    <ClInclude Include="..\vendor\lua\src\lopnames.h" />
    <ClInclude Include="..\vendor\lua\src\lparser.h" />
    <ClInclude Include="..\vendor\lua\src\lprefix.h" />
    <ClInclude Include="..\vendor\lua\src\lstate.h" />
    <ClInclude Include="..\vendor\lua\src\lstring.h" />
    <ClInclude Include="..\vendor\lua\src\ltable.h" />
    <ClInclude Include="..\vendor\lua\src\ltm.h" />
    <ClInclude Include="..\vendor\lua\src\lua.h" />
    <ClInclude Include="..\vendor\lua\src\luaconf.h" />
    <ClInclude Include="..\vendor\lua\src\lualib.h" />
    <ClInclude Include="..\vendor\lua\src\lundump.h" />
    <ClInclude Include="..\vendor\lua\src\lvm.h" />
    <ClInclude Include="..\vendor\lua\src\lzio.h" />
  </ItemGroup>
  <ItemGroup>
    <ClCompile Include="..\vendor\lua\src\lapi.c" />
    <ClCompile Include="..\vendor\lua\src\lauxlib.c" />
    <ClCompile Include="..\vendor\lua\src\lbaselib.c" />
    <ClCompile Include="..\vendor\lua\src\lcode.c" />
    <ClCompile Include="..\vendor\lua\src\lcorolib.c" />
    <ClCompile Include="..\vendor\lua\src\lctype.c" />
    <ClCompile Include="..\vendor\lua\src\ldblib.c" />
    <ClCompile Include="..\vendor\lua\src\ldebug.c" />
    <ClCompile Include="..\vendor\lua\src\ldo.c" />
    <ClCompile Include="..\vendor\lua\src\ldump.c" />
    <ClCompile Include="..\vendor\lua\src\lfunc.c" />
    <ClCompile Include="..\vendor\lua\src\lgc.c" />
    <ClCompile Include="..\vendor\lua\src\linit.c" />
    <ClCompile Include="..\vendor\lua\src\liolib.c" />
    <ClCompile Include="..\vendor\lua\src\llex.c" />
    <ClCompile Include="..\vendor\lua\src\lmathlib.c" />
    <ClCompile Include="..\vendor\lua\src\lmem.c" />
    <ClCompile Include="..\vendor\lua\src\loadlib.c" />
    <ClCompile Include="..\vendor\lua\src\lobject.c" />
    <ClCompile Include="..\vendor\lua\src\lopcodes.c" />
    <ClCompile Include="..\vendor\lua\src\loslib.c" />
    <ClCompile Include="..\vendor\lua\src\lparser.c" />
    <ClCompile Include="..\vendor\lua\src\lstate.c" />
    <ClCompile Include="..\vendor\lua\src\lstring.c" />
    <ClCompile Include="..\vendor\lua\src\lstrlib.c" />
    <ClCompile Include="..\vendor\lua\src\ltable.c" />
    <ClCompile Include="..\vendor\lua\src\ltablib.c" />
    <ClCompile Include="..\vendor\lua\src\ltm.c" />
    <ClCompile Include="..\vendor\lua\src\lua.c" />
    <ClCompile Include="..\vendor\lua\src\lundump.c" />
    <ClCompile Include="..\vendor\lua\src\lutf8lib.c" />
    <ClCompile Include="..\vendor\lua\src\lvm.c" />
    <ClCompile Include="..\vendor\lua\src\lzio.c" />
  </ItemGroup>
  ````
  
  Lua_Server.vcxproj:
```xml
  <ItemGroup>
    <ClInclude Include="..\vendor\lua\src\lapi.h" />
    <ClInclude Include="..\vendor\lua\src\lauxlib.h" />
    <ClInclude Include="..\vendor\lua\src\lcode.h" />
    <ClInclude Include="..\vendor\lua\src\lctype.h" />
    <ClInclude Include="..\vendor\lua\src\ldebug.h" />
    <ClInclude Include="..\vendor\lua\src\ldo.h" />
    <ClInclude Include="..\vendor\lua\src\lfunc.h" />
    <ClInclude Include="..\vendor\lua\src\lgc.h" />
    <ClInclude Include="..\vendor\lua\src\ljumptab.h" />
    <ClInclude Include="..\vendor\lua\src\llex.h" />
    <ClInclude Include="..\vendor\lua\src\llimits.h" />
    <ClInclude Include="..\vendor\lua\src\lmem.h" />
    <ClInclude Include="..\vendor\lua\src\lobject.h" />
    <ClInclude Include="..\vendor\lua\src\lopcodes.h" />
    <ClInclude Include="..\vendor\lua\src\lopnames.h" />
    <ClInclude Include="..\vendor\lua\src\lparser.h" />
    <ClInclude Include="..\vendor\lua\src\lprefix.h" />
    <ClInclude Include="..\vendor\lua\src\lstate.h" />
    <ClInclude Include="..\vendor\lua\src\lstring.h" />
    <ClInclude Include="..\vendor\lua\src\ltable.h" />
    <ClInclude Include="..\vendor\lua\src\ltm.h" />
    <ClInclude Include="..\vendor\lua\src\lua.h" />
    <ClInclude Include="..\vendor\lua\src\luaconf.h" />
    <ClInclude Include="..\vendor\lua\src\lualib.h" />
    <ClInclude Include="..\vendor\lua\src\lundump.h" />
    <ClInclude Include="..\vendor\lua\src\lvm.h" />
    <ClInclude Include="..\vendor\lua\src\lzio.h" />
  </ItemGroup>
  <ItemGroup>
    <ClCompile Include="..\vendor\lua\src\lapi.c" />
    <ClCompile Include="..\vendor\lua\src\lauxlib.c" />
    <ClCompile Include="..\vendor\lua\src\lbaselib.c" />
    <ClCompile Include="..\vendor\lua\src\lcode.c" />
    <ClCompile Include="..\vendor\lua\src\lcorolib.c" />
    <ClCompile Include="..\vendor\lua\src\lctype.c" />
    <ClCompile Include="..\vendor\lua\src\ldblib.c" />
    <ClCompile Include="..\vendor\lua\src\ldebug.c" />
    <ClCompile Include="..\vendor\lua\src\ldo.c" />
    <ClCompile Include="..\vendor\lua\src\ldump.c" />
    <ClCompile Include="..\vendor\lua\src\lfunc.c" />
    <ClCompile Include="..\vendor\lua\src\lgc.c" />
    <ClCompile Include="..\vendor\lua\src\linit.c" />
    <ClCompile Include="..\vendor\lua\src\liolib.c" />
    <ClCompile Include="..\vendor\lua\src\llex.c" />
    <ClCompile Include="..\vendor\lua\src\lmathlib.c" />
    <ClCompile Include="..\vendor\lua\src\lmem.c" />
    <ClCompile Include="..\vendor\lua\src\loadlib.c" />
    <ClCompile Include="..\vendor\lua\src\lobject.c" />
    <ClCompile Include="..\vendor\lua\src\lopcodes.c" />
    <ClCompile Include="..\vendor\lua\src\loslib.c" />
    <ClCompile Include="..\vendor\lua\src\lparser.c" />
    <ClCompile Include="..\vendor\lua\src\lstate.c" />
    <ClCompile Include="..\vendor\lua\src\lstring.c" />
    <ClCompile Include="..\vendor\lua\src\lstrlib.c" />
    <ClCompile Include="..\vendor\lua\src\ltable.c" />
    <ClCompile Include="..\vendor\lua\src\ltablib.c" />
    <ClCompile Include="..\vendor\lua\src\ltm.c" />
    <ClCompile Include="..\vendor\lua\src\lua.c" />
    <ClCompile Include="..\vendor\lua\src\lundump.c" />
    <ClCompile Include="..\vendor\lua\src\lutf8lib.c" />
    <ClCompile Include="..\vendor\lua\src\lvm.c" />
    <ClCompile Include="..\vendor\lua\src\lzio.c" />
  </ItemGroup>
  ```

**Comment by Lpsd** [comment]
You shouldn't need to manually edit Visual Studio project files. Once you've replaced the Lua vendor files, simply run `win-create-projects.bat` and it will update each project as necessary.

Depending on the changes made to vendor files, you may need to tweak the premake config for the vendor, as well as any projects depending on that vendor.

**Comment by Disinterpreter** [comment]
It is good, but what about compatibility? I suppose many old 5.1 scripts can be broken.  
Perhaps, you need to make a selector in meta.xml.

File: Client/mods/deathmatch/logic/lua/CLuaMain.cpp [hunk 1]
```diff
@@ -143,20 +143,14 @@ void CLuaMain::InitVM()
...
 
     // Load LUA libraries
-    luaopen_base(m_luaVM);
-    luaopen_math(m_luaVM);
-    luaopen_string(m_luaVM);
-    luaopen_table(m_luaVM);
-    luaopen_debug(m_luaVM);
-    luaopen_utf8(m_luaVM);
-    luaopen_os(m_luaVM);
+    luaL_openlibs(m_luaVM);
```
Comment (Disinterpreter) [inline Client/mods/deathmatch/logic/lua/CLuaMain.cpp:153]: Huge discussion about libs here #1972 and here https://github.com/multitheftauto/mtasa-blue/discussions/1782

**Comment by fresholia** [comment]
> Perhaps, you need to make a selector in meta.xml.

I don't think it is necessary. The problems in systems that conflict with 5.4 are more or less certain. It will be easy to solve them by Lua C.

**Comment by Lpsd** [comment]
Newer versions of Lua contain some non-backwards-compat. changes, those would need to be fixed. If not, we'd have to approach it somehow with #1591 (although I'm not even sure if that's realistic, we'd then have clients running two different Lua versions).

Our current Lua version is modified, by us. Does anyone have experience with those changes? Are they portable to 5.4?

Wouldn't our best option be to try and backport desired functionality on top of our existing modified build? On that topic - what is the desired functionality you want from Lua 5.4?

**Comment by Pirulax** [comment]
MTA has modified internal Lua stuff, so you must track the changes down, and incorporate them into this as well.

**Comment by fresholia** [comment]
> MTA has modified internal Lua stuff, so you must track the changes down, and incorporate them into this as well.

All integrated^^

**Comment by Pirulax** [comment]
I wonder if it would be possible to have 2 separate DLLs and load them on the fly, based on a value in mtaserver.conf?

**Comment by fresholia** [comment]
> Newer versions of Lua contain some non-backwards-compat. changes, those would need to be fixed. If not, we'd have to approach it somehow with #1591 (although I'm not even sure if that's realistic, we'd then have clients running two different Lua versions).

Besides two different versions of lua I have one more alternative.
There is not much difference between 5.4 and 5.1 in functions running on MTA, when we redirect the old functions to the new function on Lua C side, there will be no problem.
(I tried unpack) newest: table.unpack

**Comment by Lpsd** [comment]
> I wonder if it would be possible to have 2 separate DLLs and load them on the fly, based on a value in mtaserver.conf?

yes, this is easily achievable. it's whether we want to do that (discussion about this should go to #1591 though)

**Comment by sbx320** [comment]
Even if we made a compatibility shim to allow Lua 5.1 scripts to run with 5.4, it'd still break all compiled scripts. And supplying multiple versions of Lua doesn't seem to be worth it.

--- 
*edit*

> (discussion about this should go to #1591 though)

#1591 is about versioning **our** scripting functions, so we can change/remove them. This is a very different issue.

**Comment by cleopatradev** [comment]
> - Server crashes using exports
> - Server crashes when player join

Reason: I caught an error before the server crash, but I don't know why, any idea?
OnJoin:
Screenshot_8
Use exports:
Screenshot_7

**Comment by Allerek** [comment]
> It is good, but what about compatibility? I suppose many old 5.1 scripts can be broken.
> Perhaps, you need to make a selector in meta.xml.

As we all know, some backward incompatible functions are meant to be introduced in 1.6, in my opinion if we update Lua to newer version, it should be merged with 1.6 and doesn't care for backward compatibility, anyway, in that case we should release nightly version at least 2 months before final release, for developers to prepare their scripts. 

Thinking realistic - MTA needs some sort of backward incompatibility to make it a little bit cleaner, we have a lot of old, poorly written scripts in comunity, so backward incompatibility would maybe make people think about getting their skills a bit better.
#ChangeMyMind

**Comment by fresholia** [comment]
> in my opinion if we update Lua to newer version, it should be merged with 1.6

It's okay to adapt Lua's 5.1 functions to 5.4.
The main problem is that previously compiled resources will not work with the Lua 5.4. This will bring down a lot of servers, I'm working on this. Actually, there is no problem other than that.
I am also considering adding LuaJIT to the version selector on mtaserv.conf

**Comment by Dutchman101** [comment]
I am surprised to see a PR simply upgrading Lua, because of the amount of discussions on this subject (including in #development channel) in the past years - none of them saw it as a trivial task, and all discussions generally ended favouring backwards compatibility. 

There were some suggested implementations like mine: using a Lua 5.1 / 5.4 tag in resource meta.xml, preserving both legacy and upgraded Lua in MTA so that scripters can choose (and the resource would thus also dictate if the client loads it with 5.1 or 5.4). In that case, for a longer period of time we will probably need to keep 5.1 as default if not specified in meta.xml (which also means already compiled resources will keep working), and scripters that want to use the latest features, or who learned Lua with 5.4 and are used to it, can do so.

You can find other suggested implementations (and reasons why people believed it wasn't realistic to upgrade Lua) by locating some of the other discussions.

**Comment by Lpsd** [comment]
> I am surprised to see a PR simply upgrading Lua, because of the amount of discussions on this subject (including in #development channel) in the past years - none of them saw it as a trivial task

The hard part is yet to be done. All this does is upgrade vendor files. Many issues remain.

**Comment by fresholia** [comment]
> There were some suggested implementations like mine: using a Lua 5.1 / 5.4 tag in resource meta.xml, preserving both legacy and upgraded Lua in MTA so that scripters can choose (and the resource would thus also dictate if the client loads it with 5.1 or 5.4). In that case, for a longer period of time we will probably need to keep 5.1 as default if not specified in meta.xml (which also means already compiled resources will keep working), and scripters that want to use the latest features, or who learned Lua with 5.4 and are used to it, can do so.

Yes, I agree. I said the same idea on the development channel. They did not find it very logical to run two Lua VMs. Actually, I have overcome many problems. I guess the only problem I couldn't deal with was reading the backward luac files in the new VM.

**Comment by se16n** [comment]
Hello everyone,

I'm wondering why Lua needs to be updated to the 5.4? Most of the world is de-facto stick to 5.1 so scripts are.

If there's a change involving compiled scripts breakage then it's might worth to consider replace vanilla Lua 5.1 with LuaJIT (interpreter mode which is still significantly faster than vanilla).

**Comment by ghost** [comment]
@se16n see [this](https://github.com/multitheftauto/mtasa-blue/discussions/903) for discussion about LuaJIT.

**Comment by fresholia** [comment]
I added a version selector to save backward and compiled files.

Since the default Lua version is 5.1, this feature will only be available to developers as an option.

**Comment by Pirulax** [comment]
MTA's biggest mistake was to allow precompiled files, or well... guarantee that they will work.
Is it possible to write a 5.1 to 5.4 transpiler? As of my knowledge they added only 2 new instructions.

**Comment by fresholia** [comment]
invalid

**Pull request closed without merging**